Sources/PenguinStructures/Projections.swift

```diff
@@ -0,0 +1,169 @@
+  public var endIndex: Index { base.endIndex }
+
+  /// The indices of all elements, in order.
+  public var indices: Indices { base.indices }
+
+  /// The position following `x`.
+  public func index(after x: Index) -> Index {
+    base.index(after: x)
+  }
+
+  /// Replaces `x` with its successor
+  public func formIndex(after x: inout Index) {
+    base.formIndex(after: &x)
+  }
+
+  /// Accesses the element at `i`.
+  public subscript(i: Index) -> Element {
+    get { base[i][keyPath: BaseElementPart.focus] }
+  }
+
+  /// True iff `self` contains no elements.
+  public var isEmpty: Bool { base.isEmpty }
+
+  /// The number of elements in `self`.
+  ///
+  /// - Complexity: O(1) if `Base` conforms to `RandomAccessCollection`; otherwise, O(N) where N is
+  ///   the number of elements.
+  public var count: Int { base.count }
+
+  /// Returns an index that is the specified distance from the given index.
+  ///
+  /// - Complexity: O(1) if `Base` conforms to `RandomAccessCollection`; otherwise, O(`distance`).
+  public func index(_ i: Index, offsetBy distance: Int) -> Index {
+    base.index(i, offsetBy: distance)
+  }
+
+  /// Returns `i` offset forward by `distance`, unless that distance is beyond a given limiting
+  /// index, in which case nil is returned.
+  ///
+  /// - Complexity: O(1) if `Base` conforms to `RandomAccessCollection`; otherwise, O(`distance`).
+  public func index(
+    _ i: Index, offsetBy distance: Int, limitedBy limit: Index
+  ) -> Index? {
+    base.index(i, offsetBy: distance, limitedBy: limit)
+  }
+
+  /// Returns the number of positions between `start` to `end`.
+  ///
+  /// - Complexity: O(1) if `Base` conforms to `RandomAccessCollection`; otherwise, worst case
+  ///   O(`count`).
+  public func distance(from start: Index, to end: Index) -> Int {
+    base.distance(from: start, to: end)
+  }
+}
+
+extension Projections: MutableCollection
+  where Base: MutableCollection,
+        BaseElementPart.Focus : WritableKeyPath<Base.Element, BaseElementPart.Value>
+{
+  /// Accesses the element at `i`.
+  public subscript(i: Index) -> Element {
+    get { base[i][keyPath: BaseElementPart.focus] }
+    set { base[i][keyPath: BaseElementPart.focus] = newValue }
+    _modify { yield &base[i][keyPath: BaseElementPart.focus] }
+  }
+}
+
+extension Projections: BidirectionalCollection
+  where Base: BidirectionalCollection
+{
+  /// Returns the position immediately before `i`.
+  public func index(before i: Index) -> Index {
+    return base.index(before: i)
+  }
+  
+  /// Replaces the value of `i` with its predecessor.
+  public func formIndex(before i: inout Index) {
+    return base.formIndex(before: &i)
+  }
+}
+
+extension Projections: RandomAccessCollection
+  where Base: RandomAccessCollection
+{}
+
+extension Sequence {
+  /// Accesses a sequence consisting of the elements of `self` projected through the given lens.
+  public subscript<L: Lens>(lens _: Type<L>) -> Projections<Self, L>
```

Comment:
Is subscript the right expression of this approach? (Concretely, collection transformations like `map` are functions instead of subscripts. Perhaps subscript only makes sense for the mutable collection where there's write-back?)

Reply:
Subscript is certainly *required* for the mutable collection where there's write-back.  There's just no alternative.  Given that we have both read-only and read-write subscripts, I'd think giving the mutable and immutable cases nonuniform syntax would require some significant justification, no?